### PR TITLE
Add velocity array support for trajectories scan

### PR DESCRIPTION
**Pull request by LeandroMartinsdS**
This PR is related to the working in progress to remove the velocities calculation from the motion program, and bring it to the driver level.

Some points that still need to be addressed, some of them discussed with @coretl , @ajgdls  and  @JimboMonkey :

- [x] Fix white spaces to make it clear where are the changes
- [x] Implement the velocities calculation in the driver level instead of in motion program
  - [x] Rollback appendVelMode/appendVelArray to append
  - [x] Substitute VelType for CalcVel, and remove from the broker
  - [x] Set CalcVel using asyn
  - [x] Port the equations from the motion program
- [x] Stop sending the VelMode
- [x] Update motion programs removing the velocities calculations (and correspondent subprograms)
- [x] Update the memory buffer
- [x] Bump motion program version
- [x] Update tests

If I'm missing something, please let me know.

### CODE REVIEW

**Comment by ajgdls** [comment]
Hi @LeandroMartinsdS I agree with your bullet list of tasks to complete this implementation.

**Comment by ajgdls** [comment]
Hi @LeandroMartinsdS at the moment it is not possible to download the motion programs to the PMAC.  I think there are a few syntax errors in the pmc files.  I would recommend the following checks:
- Remove any commented lines (I notice a few where there is a "//" but not a ";") to prevent download errors
- There are still instances of using variables (e.g. Prev_A) but these are no longer defined in the script as they have been commented out.

File: pmacApp/pmc/trajectory_scan_clipper.pmc [hunk 1]
```diff
@@ -2,7 +2,7 @@
 ; Set these values for your PMAC
 ; *****************************************************************************************
 #define VarAdr       B8     ; Prefix of address (eg VarAdr=300 then $30000 is the start address)
-#define BuffLen      950    ; Length of buffers
+#define BuffLen      500    ; Length of buffers
```
Comment (ajgdls) [inline pmacApp/pmc/trajectory_scan_clipper.pmc:5]: The amount of required memory has increased by close to a factor of two, and this buffer length number has decreased but not by factor of two.  Have you checked this fits within the allowed memory space?

File: pmacApp/pmc/trajectory_scan_code_ppmac.pmc [hunk 2]
```diff
@@ -137,57 +137,48 @@ N102:
...
         Current_V = Next_V(CalculatedBase)
-        V_Prev_Vel = V_Vel
+        V_Vel = Next_V_Vel(CalculatedBase)
     }
     If (W_Axis == 1)
     {
-        Prev_W = Current_W
         Current_W = Next_W(CalculatedBase)
-        W_Prev_Vel = W_Vel
+        V_Vel = Next_V_Vel(CalculatedBase)
```
Comment (ajgdls) [inline pmacApp/pmc/trajectory_scan_code_ppmac.pmc]: This needs to map W velocities

File: pmacApp/src/pmacController.cpp [hunk 3]
```diff
@@ -726,6 +738,7 @@ void pmacController::setupBrokerVariables(void) {
   pBroker_->addReadVariable(pmacMessageBroker::PMAC_SLOW_READ, PMAC_TRAJ_BUFFER_LENGTH);
   pBroker_->addReadVariable(pmacMessageBroker::PMAC_SLOW_READ, PMAC_TRAJ_BUFF_ADR_A);
   pBroker_->addReadVariable(pmacMessageBroker::PMAC_SLOW_READ, PMAC_TRAJ_BUFF_ADR_B);
+  // pBroker_->addReadVariable(pmacMessageBroker::PMAC_SLOW_READ, PMAC_TRAJ_VELOCITY_TYPE);
```
Comment (ajgdls) [inline pmacApp/src/pmacController.cpp]: Added but then left commented out, is this required?

File: pmacApp/src/pmacController.cpp [hunk 4]
```diff
@@ -1350,7 +1368,7 @@ asynStatus pmacController::mediumUpdate(pmacCommandStore *sPtr) {
           } else {
             gpioOutputs += gpioBit << gpio;
           }
-        }
+         }
```
Comment (ajgdls) [inline pmacApp/src/pmacController.cpp]: Remove incorrect whitespace

File: pmacApp/src/pmacController.cpp [hunk 5]
```diff
@@ -2740,6 +2790,7 @@ asynStatus pmacController::buildProfile(int csNo) {
 
     // Check for any invalid times
     int maxValue = 0xFFFFFF;
+    // TODO:  Add cid check for pppmac
```
Comment (ajgdls) [inline pmacApp/src/pmacController.cpp]: Has this been completed?

File: pmacApp/src/pmacController.cpp [hunk 6]
```diff
@@ -2781,10 +2832,10 @@ asynStatus pmacController::buildProfile(int csNo) {
   if (status == asynSuccess) {
     // Initialise the trajectory store
     status = pTrajectory_->initialise(numPoints);
-
+
```
Comment (ajgdls) [inline pmacApp/src/pmacController.cpp]: Whitespace

File: pmacApp/src/pmacController.cpp [hunk 7]
```diff
@@ -4486,6 +4625,29 @@ asynStatus pmacMonitorVariables(const char *controller, const char *variablesStr
...
+//   if (!pC)
+//   {
+//     printf("%s:%s: Error port %s not found\n",
+//            driverName, functionName, controller);
+//     return asynError;
+//   }
+
+//   return pC->pmacSetTrajectoryUseVelArray();
+// }
+
```
Comment (ajgdls) [inline pmacApp/src/pmacController.cpp]: Remove this commented code if it is not required?

File: pmacApp/unitTests/test_PMACTrajectory.cpp [hunk 8]
```diff
@@ -86,14 +96,19 @@ BOOST_AUTO_TEST_CASE(test_PMACTrajectory)
   int userVal;
   BOOST_CHECK_EQUAL(trajectory.getUserMode(-1, &userVal), asynError);
   BOOST_CHECK_EQUAL(trajectory.getUserMode(20, &userVal), asynError);
-  int velocityVal;
-  BOOST_CHECK_EQUAL(trajectory.getVelocityMode(-1, &velocityVal), asynError);
-  BOOST_CHECK_EQUAL(trajectory.getVelocityMode(20, &velocityVal), asynError);
+  // int velocityVal;
+  // BOOST_CHECK_EQUAL(trajectory.getVelocityMode(-1, &velocityVal), asynError);
+  // BOOST_CHECK_EQUAL(trajectory.getVelocityMode(20, &velocityVal), asynError);
```
Comment (ajgdls) [inline pmacApp/unitTests/test_PMACTrajectory.cpp]: Remove commented code

File: pmacApp/unitTests/test_PMACTrajectory.cpp [hunk 9]
```diff
@@ -104,18 +119,23 @@ BOOST_AUTO_TEST_CASE(test_PMACTrajectory)
...
   BOOST_CHECK_EQUAL(trajectory.getUserMode(19, &userVal), asynSuccess);
   BOOST_CHECK_EQUAL(userVal, 1);
-  BOOST_CHECK_EQUAL(trajectory.getVelocityMode(4, &velocityVal), asynSuccess);
-  BOOST_CHECK_EQUAL(velocityVal, 5);
-  BOOST_CHECK_EQUAL(trajectory.getVelocityMode(14, &velocityVal), asynSuccess);
-  BOOST_CHECK_EQUAL(velocityVal, 5);
+  // BOOST_CHECK_EQUAL(trajectory.getVelocityMode(4, &velocityVal), asynSuccess);
+  // BOOST_CHECK_EQUAL(velocityVal, 5);
+  // BOOST_CHECK_EQUAL(trajectory.getVelocityMode(14, &velocityVal), asynSuccess);
+  // BOOST_CHECK_EQUAL(velocityVal, 5);
```
Comment (ajgdls) [inline pmacApp/unitTests/test_PMACTrajectory.cpp]: Remove commented code

File: pmacApp/unitTests/test_PMACTrajectory.cpp [hunk 10]
```diff
@@ -124,7 +144,8 @@ BOOST_AUTO_TEST_CASE(test_PMACTrajectory)
   BOOST_CHECK_EQUAL(trajectory.getTotalNoOfPoints(), 10000);
 
   // Try to append once more and verify it fails
-  BOOST_CHECK_EQUAL(trajectory.append(pos, time, user, vel, 10), asynError);
+  // BOOST_CHECK_EQUAL(trajectory.append(pos, time, user, velMode, 10), asynError);
```
Comment (ajgdls) [inline pmacApp/unitTests/test_PMACTrajectory.cpp]: Remove commented code

**Review by ajgdls — CHANGES_REQUESTED** [review]

**Comment by ajgdls** [comment]
Hi @LeandroMartinsdS mostly minor comments, just the velocity mapping that looks incorrect (V and W mixed).

File: pmacApp/pmc/trajectory_scan_clipper.pmc (hunk 1 above)
Comment (LeandroMartinsdS) [reply pmacApp/pmc/trajectory_scan_clipper.pmc:5]: The number of buffers had increased from 10 to 19 by adding the velocities for each axis.
The buffer length has decreased applying the ratio of 10/19, resulting in the buffer length of 500.
By my tests, the buffer fits even for a value close to 530, but supposing that already has some conservative margin I keep the value of 500.

Comment (ajgdls) [reply pmacApp/pmc/trajectory_scan_clipper.pmc:5]: OK perfect.

**Comment by LeandroMartinsdS** [comment]
@ajgdls changes are all pushed.

**Review by ajgdls — APPROVED** [review]

**Pull request merged**